x-pack/plugins/rule_registry/server/utils/create_lifecycle_executor.ts

```diff
@@ -228,59 +227,77 @@ export const createLifecycleExecutor = (
-      ...alertData,
-      ...commonRuleFields,
-      [ALERT_DURATION]: (options.startedAt.getTime() - new Date(started).getTime()) * 1000,
-      [ALERT_INSTANCE_ID]: alertId,
-      [ALERT_START]: started,
-      [ALERT_STATUS]: isActive ? ALERT_STATUS_ACTIVE : ALERT_STATUS_RECOVERED,
-      [ALERT_WORKFLOW_STATUS]: alertData[ALERT_WORKFLOW_STATUS] ?? 'open',
-      [ALERT_UUID]: alertUuid,
-      [EVENT_KIND]: 'signal',
-      [EVENT_ACTION]: isNew ? 'open' : isActive ? 'active' : 'close',
-      [VERSION]: ruleDataClient.kibanaVersion,
-      ...(isRecovered ? { [ALERT_END]: commonRuleFields[TIMESTAMP] } : {}),
-    };
-
-    return event;
-  });
+  const makeEventsDataMapFor = (alertIds: string[]) =>
+    alertIds.map((alertId) => {
+      const alertData = trackedAlertsDataMap[alertId];
+      const currentAlertData = currentAlerts[alertId];
+
+      if (!alertData) {
+        logger.warn(`Could not find alert data for ${alertId}`);
+      }
+
+      const isNew = !state.trackedAlerts[alertId];
+      const isRecovered = !currentAlerts[alertId];
+      const isActive = !isRecovered;
+
+      const { alertUuid, started } = state.trackedAlerts[alertId] ?? {
+        alertUuid: v4(),
+        started: commonRuleFields[TIMESTAMP],
+      };
+
+      const event: ParsedTechnicalFields = {
+        ...alertData?.fields,
+        ...commonRuleFields,
+        ...currentAlertData,
+        [ALERT_DURATION]: (options.startedAt.getTime() - new Date(started).getTime()) * 1000,
+
+        [ALERT_INSTANCE_ID]: alertId,
+        [ALERT_START]: started,
+        [ALERT_UUID]: alertUuid,
+        [ALERT_STATUS]: isRecovered ? ALERT_STATUS_RECOVERED : ALERT_STATUS_ACTIVE,
+        [ALERT_WORKFLOW_STATUS]: alertData?.fields[ALERT_WORKFLOW_STATUS] ?? 'open',
+        [EVENT_KIND]: 'signal',
+        [EVENT_ACTION]: isNew ? 'open' : isActive ? 'active' : 'close',
+        [VERSION]: ruleDataClient.kibanaVersion,
+        ...(isRecovered ? { [ALERT_END]: commonRuleFields[TIMESTAMP] } : {}),
+      };
+
+      return {
+        indexName: alertData?.indexName,
+        event,
+      };
+    });
+
+  const trackedEventsToIndex = makeEventsDataMapFor(trackedAlertIds);
+  const newEventsToIndex = makeEventsDataMapFor(newAlertIds);
+  const allEventsToIndex = [...trackedEventsToIndex, ...newEventsToIndex];
 
-  if (eventsToIndex.length > 0 && ruleDataClient.isWriteEnabled()) {
-    logger.debug(`Preparing to index ${eventsToIndex.length} alerts.`);
+  if (allEventsToIndex.length > 0 && ruleDataClient.isWriteEnabled()) {
+    logger.debug(`Preparing to index ${allEventsToIndex.length} alerts.`);
 
     await ruleDataClient.getWriter().bulk({
-      body: eventsToIndex.flatMap((event) => [{ index: { _id: event[ALERT_UUID]! } }, event]),
+      body: allEventsToIndex.flatMap(({ event, indexName }) => [
+        indexName
+          ? { index: { _id: event[ALERT_UUID]!, _index: indexName, require_alias: false } }
+          : { index: { _id: event[ALERT_UUID]! } },
```

Comment:
This looks good to me, I'd just like to ask if you thought about using `create` + `update` instead of `index` and decided to keep `index` for both operations. It looks like this executor collects the full set of document fields, so it's safer to use `index` (create or replace the whole doc).

Reply:
Yes, we considered `update` but figured using `index` gives tighter control over the full document content (e.g. allows for removal of fields). We might refactor to not fetch the full content and use `update` in the future, but it looked like too much of a change for a late 7.15.0 bug-fix.